juju/juju
# Address/Space Re-Model; Address Spaces Stored as IDs

## Description of change

This patch changes addresses to be stored in state with a space ID instead of a space name and provider ID.

The `core/network` package has the following notable changes:
- `Address` is now an indirection for concrete types:
  - `MachineAddress` for space-unaware addresses.
  - `ProviderAddress` for addresses supplied by providers or space-aware addresses resolved from state. This type has the same fields as the old `Address` type.
  - `SpaceAddress` for working server-side with space-aware addresses, decorated with a space ID.
- `HostPort` is now an indirection for concrete types congruent with addresses above:
  - `MachineHostPort`.
  - `ProviderHostPort`.
  - `SpaceHostPort`.
- Collections of these types are generally declared as new types, with transformation methods to return collections of the indirections. Common methods reside on the indirected collection type where required.

In the state package, the `Address` type corresponds with `network.SpaceAddress` (stored with an optional space ID). Transformations to/from the `core/network` type are adjusted accordingly. This change affects many modules from backups, application/machine/unit, migrations, relations etc.

Methods that used to do specific filtering/ordering according scope have been generalised to accept a scope matching function, with the choice of matcher now a concern of the caller.

Patches will follow to handle migrations and upgrades.

_Note for reviewers: Most of the patch is symbol changes for the new types. I have attempted to add comments to pertinent parts._

## QA steps

All of these need to be accompanied by log checks for any errors.
- MAAS (done by author)
  - Bootstrap and ensure that spaces are loaded.
  - Set management and HA spaces.
  - Enable HA and ensure success.
  - Check agent configurations for the correct API address based on management space.
  - Deploy some units.
- AWS (done by author)
  - Bootstrap and ensure spaces can be added.
  - Deploy and relate some charms using one of the added spaces for endpoint bindings.
- Openstack (done by author)
  - Deploy kubernetes-core.
- LXD (done by author)
  - Deploy webscale test bundle.
- Azure, K8s (done by wallyworld)
  - Deploy PostreSQL.
  - Add some units.
- Vsphere (done by babbageclunk)
  - Deploy PostreSQL.
  - Add some units.
- General
  - Ensure `juju ssh` works.

## Documentation changes

None.

## Bug reference

N/A
